kingdom/access/authorization/verify.py

```diff
@@ -0,0 +1,181 @@
+    Consider a subject that tries to access a resource. The access attempt is
+    abstracted as AccessRequest and the whole set of policies it owns is
+    represented as PolicyContext.
+
+    This function returns whether the user has enough permissions to do
+    the write operation on AccessRequest.
+
+    >>> access_request = AccessRequest(
+        operation=Permission.CREATE, resource=Resource("Coupon"), selector="*"
+    )
+    >>> owned_perm = {
+        "coupon": {
+            "ab4c": (Permission.READ,),
+            "bc3f": (Permission.READ, Permission.UPDATE),
+            "cc4a": (Permission.UPDATE,),
+            "b4a3": (Permission.DELETE,),
+        },
+    }
+    >>> check_write_permission(owned_perm, access_request)
+    False
+
+    More examples on test suite.
+    """
+    # TODO: Refactor and improve logic handling, DRY plz
+    # Sanity check.
+    assert access_request.operation & (
+        Permission.CREATE | Permission.UPDATE | Permission.DELETE
+    )
+
+    resource = access_request.resource
+    if resource not in owned_policies:
+        # Resource is unknown to subject.
+        return False
+    # resource_owned = owned_policies[resource]
+
+    # CREATE case is different than UPDATE and DELETE.
+    # It requires a "*" selector.
+    if access_request.operation == Permission.CREATE.value:
+        # To CREATE, subject must have at least one '*' policy.
+        if TOKEN_ALL not in owned_policies[resource]:
+            return False
+
+        # Checking is a simple O(1) step.
+        return is_allowed(
+            owned_policies[resource][TOKEN_ALL], access_request.operation
+        )
+        # return Permission.CREATE in owned_policies[resource][TOKEN_ALL]
+
+    # Deal with UPDATE or DELETE
+    # Check for generics.
+    if TOKEN_ALL in owned_policies[resource]:
+        permissions = owned_policies[resource][TOKEN_ALL]
+        if is_allowed(permissions, access_request.operation):
+            # Wildcard matches permission.
+            return True
+
+    # Then specify.
+    if access_request.selector not in owned_policies[resource]:
+        # No rule for requested instance. Denied.
+        return False
+
+    permissions = owned_policies[resource][access_request.selector]
+    return is_allowed(permissions, access_request.operation)
+
+
+def is_allowed(
+    owned_permissions: Union[Tuple[Permission, ...], int],
+    requested_operation: Union[int, Permission],
+) -> bool:
+    """
+    When a subject tries to perform a write, this function calculates whether
+    the requested operation is allowed on a set of owned permissions.
+
+    There are a few caveats:
+     1. Write permission overrides read permission.
+        e.g. If a Subject has any write permission, him/her implicitly
+        have read permission.
+     2. No write permissions override one-another.
+        e.g. If a Subject has DELETE permission and it doesn't have
+        UPDATE or CREATE permissions, it is only allowed to delete.
```

Comment:
this clarifies a lot about how this permissions are working, i don't think this premise was exposed in the issue for this

Reply:
but also, i have a doubt about write permissions overriding read, specifically on CREATE operation, for a user to be allowed to perform this operation it need a `*` selector, so a user that is capable of create a resource is capable of reading all resources, its that expected?

Reply:
You are totally right! I do have to update the issue when this got merged for future reference.

Reply:
@rafamelos Yes, this implicit behaviour is the only one expected in all of those permissions. You think that's reasonable? I mean, if a user is able to create new instances of a resource, it should be able to see them all. Or do you think and can see this as a limitation?

Reply:
I think it's okay by now.